### PR TITLE
Pymbar 4 compatibility

### PR DESCRIPTION
**Pull request by mattwthompson**
Splitting out some of #259 to tackle these updates one at a time

### CODE REVIEW

File: src/liquid.py [hunk 1]
```diff
@@ -935,8 +935,8 @@ def fill_weights(weights, phase_points, mbar_points, snapshots):
                         mU_kln[k, m, :]  = mE[mE_idx]
                         mU_kln[k, m, :] *= beta
                 if np.abs(np.std(mE)) > 1e-6 and mBSims > 1:
-                    mmbar = pymbar.MBAR(mU_kln, mN_k, verbose=False, relative_tolerance=5.0e-8, method='self-consistent-iteration')
-                    mW1 = mmbar.getWeights()
+                    mmbar = pymbar.MBAR(mU_kln, mN_k, verbose=False, relative_tolerance=5.0e-8)
+                    mW1 = mmbar.weights()
```
[the remark this answers is not in the document]
Comment (mattwthompson) [reply src/liquid.py:939]: Removing the argument gets different/bad behavior.

I filed a ticket with the pymbar developers: https://github.com/notifications

**Comment by leeping** [comment]
Hi Matt,

Thanks a lot for your help on this.  One of my concerns was that the major use cases for pymbar involved running a large number of long simulations, which could not be covered in the unit tests out of necessity.  (About 50 simulations of water boxes at different temperatures with 8 ns each).  It might be possible to design a unit test by storing a large 3D matrix of energies extracted from one of these jobs.  Do you think I could add this as a test, with the expected result provided by pymbar 2, and make the result a requirement for proceeding with the newer versions?

- Lee-Ping

**Comment by mattwthompson** [comment]
Sure! My objective here is to update the code to work with the new API, and improved tests would help there. If you write up that test that works with version 3 in a separate PR I can incorporate it into here - I think that would be the path with the minimum friction for you?

That the test is running and getting to the assertion step indicates to me that these changes might be correct - but maybe the method/solver protocol is misbehaving because of insufficient simulation data? I'm a bit over my head in really understanding what the existing test does.

**Comment by leeping** [comment]
Hi Matt,

I just took a closer look at the tests, and I think that it should be a good enough to determine the correctness of pymbar 4.

The test of pymbar happens in the folder `src/tests/files/test_liquid`.  A single point FB calculation is run in this folder, and it reads stored simulation results from the folder `single.tmp` (there are 6 different simulations with 80 energies each).  The MBAR calculation is based on these simulation results, and the input should be a 6 x 6 x 80 tensor.

In short, I don't think we need to add another test.  However, it is worth thinking about whether we could check for regressions in performance, in terms of wall time or number of cycles taken to reach convergence.  If I remember correctly, pymbar 3 was slower than pymbar 2 because some C++ helper code was removed, but the slowdown was still acceptable.  If the default convergence algorithm is changed in pymbar 4, then the number of iterations required to converge might increase as well.  I'd like to avoid a future situation where a future user finds out that pymbar took ~1 hour to complete a step that took minutes several years ago.

- Lee-Ping

**Comment by mattwthompson** [comment]
Hi @leeping,

I agree with your concerns about performance but I think it's something we should tackle in a future PR, for a few reasons:
* It's not obvious what could be done in the source code here to keep the MBAR-reliant paths performant and in order to report regressions to the `pymbar` developers we'd probably want a some sort of standalone test or set of benchmarks, not something tied into in the test suite here
* This at least gets things compatible with the most recent version (assuming the macOS builds don't surprise me) which is an improvement over needing to pin to old versions, even if there is a slowdown

In principle I could add some sort of a check to make sure the water study test takes less than some time to run, but I think that's undesirable because it doesn't really test the code here and is prone to flakiness due to the number of factors (particularly upstream) that can cause performance to go up or down.

**Pull request closed without merging**